kubernetes/config/kube_config_test.py

```diff
@@ -0,0 +1,413 @@
+            TEST_FILE_KEY: self._create_temp_file(
+                content=TEST_ANOTHER_DATA)}
+        t = FileOrData(obj=obj, file_key_name=TEST_FILE_KEY,
+                       data_key_name=TEST_DATA_KEY)
+        self.assertEqual(TEST_DATA_BASE64, t.data)
+
+    def test_file_given_file_and_data(self):
+        obj = {
+            TEST_DATA_KEY: TEST_DATA_BASE64,
+            TEST_FILE_KEY: self._create_temp_file(
+                content=TEST_ANOTHER_DATA)}
+        t = FileOrData(obj=obj, file_key_name=TEST_FILE_KEY,
+                       data_key_name=TEST_DATA_KEY)
+        self.assertEqual(TEST_DATA, self.get_file_content(t.file))
+
+
+class TestConfigNode(TestRoot):
+
+    test_obj = {"key1": "test", "key2": ["a", "b", "c"],
+                "key3": {"inner_key": "inner_value"},
+                "with_names": [{"name": "test_name", "value": "test_value"},
+                               {"name": "test_name2",
+                                "value": {"key1", "test"}},
+                               {"name": "test_name3", "value": [1, 2, 3]}]}
+
+    def test_normal_map_array_operations(self):
+        node = ConfigNode("test_obj", self.test_obj)
+        self.assertEqual("test", node['key1'])
+        self.assertEqual(4, len(node))
+
+        self.assertEqual("test_obj/key2", node['key2'].name)
+        self.assertEqual(["a", "b", "c"], node['key2'].value)
+        self.assertEqual("b", node['key2'][1])
+        self.assertEqual(3, len(node['key2']))
+
+        self.assertEqual("test_obj/key3", node['key3'].name)
+        self.assertEqual({"inner_key": "inner_value"}, node['key3'].value)
+        self.assertEqual("inner_value", node['key3']["inner_key"])
+        self.assertEqual(1, len(node['key3']))
+
+    def test_get_with_name(self):
+        node = ConfigNode("test_obj_with_names",
+                          self.test_obj["with_names"])
+        self.assertEqual(
+            "test_value",
+            node.get_with_name("test_name")["value"])
+        self.assertTrue(
+            isinstance(node.get_with_name("test_name2"), ConfigNode))
+        self.assertTrue(
+            isinstance(node.get_with_name("test_name3"), ConfigNode))
+        self.assertEqual("test_obj_with_names[name=test_name2]",
+                         node.get_with_name("test_name2").name)
+        self.assertEqual("test_obj_with_names[name=test_name3]",
+                         node.get_with_name("test_name3").name)
+
+    def expect_exception(self, func, message_part):
+        try:
+            func()
+            self.fail("Should fail.")
+        except ConfigException as e:
+            self.assertTrue(
+                message_part in e.message, "'%s' should be in '%s'" %
+                (message_part, e.message))
+
+    def test_invalid_key(self):
+        node = ConfigNode("test_obj", self.test_obj)
+        self.expect_exception(lambda: node['non-existance-key'],
+                              "Expect key non-existance-key in test_obj")
```

Comment:
Duplicating error messages in testing makes for brittle tests.  If testing is comprehensive, it should be enough to isolate the failure condition:

```python
def setUp(self):
  self.node = ConfigNode("test_obj", self.test_obj)

def test_get_with_name_fails_if_value_not_a_list(self):
  self.node._value = None
  with self.assertRaises(ConfigException):
    self.node.get_with_name('')